pkg/joinserver/joinserver.go

```diff
@@ -333,6 +345,13 @@ func (js *JoinServer) HandleJoin(ctx context.Context, req *ttnpb.JoinRequest, au
 					dev.UsedDevNonces = append(dev.UsedDevNonces, 0)
 					copy(dev.UsedDevNonces[i+1:], dev.UsedDevNonces[i:])
 					dev.UsedDevNonces[i] = dn
+
+					if len(dev.UsedDevNonces) > js.devNonceLimit {
+						dev.UsedDevNonces = dev.UsedDevNonces[1:]
```

Comment:
The behavior here is a bit weird, or rather, unexpected. Evicting the oldest nonce like this will actually evict the lowest nonce, not necessarily the oldest one. I think that maybe we should drop the sorting.

On a separate note: Why is the `i == 0` comparison needed ? `limit == 0` is not allowed per validation

Reply:
Yeah the weird behaviour is the biggest problem with keeping the array sorted, I could remove it but I'm wondering if the the impact of leaving like this outweighs the benefit of being a lot faster in situations with higher limits.

The `i == 0` refers to the result of the quicksort, which in this section of the code implies that it's different than the smallest value that was previously stored.

Reply:
Unresolving this comment, although it looks like the code changed since your earlier discussion.

I have two comments here:
1. Indeed there is no need to sort. But it is more likely that the last DevNonce are reused than that the first DevNonces are reused. So on L344, loop the used DevNonces backwards and keep things unsorted
2. Since you have `[1:]` here, you basically pop the first one only. But what happens if the limit was 100 and the operator reconfigured to a limit of 10? Since you only pop 1 every time, the number of DevNonces stored will stay 100. See suggestion below